### DISCUSSION TITLE
Set up repository mirror on Codeberg

### DISCUSSION THREAD

**Discussion by erkinalp** [darkreader/darkreader#10749] [General]
Move this repository to codeberg which is an instance of forgejo, which is a Gitea fork. This would mean one less proprietary web service to rely on.

**Comment by Gusted** [comment]
That's not my call to make (I'm obviously in favor of this move). CC @alexanderby

**Reply by Gusted** [reply]
For what's worth, I made [a placeholder organization](https://codeberg.org/org/darkreader) to avoid being name sniped.

**Reply by bershanskiy** [reply]
That organization is private, right? I get a 404 when I attempt to open that page. Also, I just created the account at https://codeberg.org/bershanskiy

**Reply by Gusted** [reply]
Ah sorry, I should have [linked this](https://codeberg.org/darkreader) instead.

**Comment by bershanskiy** [comment]
This is a decision for @alexanderby to make and we would not make a jump right away. However, I would be glad to discuss a plan and collaborate on a prototype for the new workflows.

Everything below is my personal subjective opinion:
1. I believe we should avoid vendor lock-in, including into GitHub ecosystem. git is inherently decentralized so at any point we can re-upload the entire repo to another service.
2. git hosting is only a fraction of the services GitHub provides. We also depend on GitHub Actions for CI and GitHub social ecosystem (network).
3. GitHub is probably the least toxic example of "Big Tech" company there is. As of now, nothing warrants immediate move away from GitHub.
4. Any GitHub user deeply relies on credential management provided by GitHub (usually without even realizing this). We are not a exception.

Therefore we could:
1. set up mirror(s) of git content on other services. These can be read-only mirrors to start with. If need be, we can use any combination of:
 - Codeberg https://codeberg.org/org/darkreader
 - GitLab https://gitlab.com/darkreader
 - Gitea - No org yet
2. avoid GitHub-specific attributes in git history
 - avoid references to issues and PRs in git history: instead of commit titles like "Fix for example.org (#123456)" have titles like "Fix for example.com" and a comment under the commit linking to the PR
3. avoid GitHub Actions and GitHub Pages lock-in
 - look into running GitHub Actions CI outside of GitHub. Ideally, this should be a drop-in replacement utilizing the same GitHub Actions syntax
4. Look into signing our commits with digital signatures besides the ones provided by GitHub. Ideally, these keys should be stored on special devices called [hardware security modules](https://en.wikipedia.org/wiki/Hardware_security_module). This is for security, of course, and certainly not because people tend to loose USB drives, break their SSDs/hard drives, etc. :)

@erkinalp If you are interested in tackling any of these challenges, I would be glad to discuss specifics.

**Reply by erkinalp** [reply]
1 and 2 are easily doable. 3 and 4 are a lot harder.

**Reply by bershanskiy** [reply]
Yes, that's exactly the point I'm truing to make. It's trivial to cyber-squat organization called `darkreader` on a bunch of git hosting services and to upload repository there once. It's a relatively easy to keep all mirrors in sync (especially if we can tolerate some delay). It's really hard to actually "move" to any of these services in any real sense and then convince everyone else to follow.

**Reply by bershanskiy** [reply]
Speaking of 3: I see they have Codeberg Pages (roughly equivalent to GitHub Pages), and some CI offering. Do you happen to know how they compare?

About 4: does Codeberg have commit signing service like GitHub? I can make something roughly equivalent on CI, but that still would require a fair bit of time. E.g., have a "main" branch like here, but then also have an ephemeral "staging" branch: all PRs are filed against "staging", then on every commit/merge to "staging" CI tests the code, then rebases and squashes the changes into a single commit, then signs the commit and pushes it to "main", and resets "staging" branch by force-pushing the new "main" head to it.

**Reply by Gusted** [reply]
> Gitea - No org yet

Gitea.com is not a production server, I think just Codeberg+Gitlab would be the safest and sanest choice without having too much mirrors.

![image](https://user-images.githubusercontent.com/25481501/218340212-596af87a-547c-4147-a043-310b8d1e8f56.png)

> Look into running GitHub Actions CI outside of GitHub. Ideally, this should be a drop-in replacement utilizing the same GitHub Actions syntax

There's [woodpecker](https://github.com/woodpecker-ci/woodpecker) and [act](https://github.com/nektos/act), that we can utilize.

**Comment by bershanskiy** [comment]
I'm very skeptical about commit signing feature they have:
1. The public key database seems to be rather sparse because it's missing even the GitHub official key
2. They do not offer to sign commits during merge (GitHub signs commits automatically)
3. There is supposedly a way to add your own codesigning key, but I failed to find UI responsible for this feature

**Comment by alexanderby** [comment]
Hi all!

So I think it's great to have a mirror on Codeberg. Thanks for taking the name from squatters. I can even set up some monthly payment for the storage service if necessary. Hope the issues could also be copied, but in a readonly mode, to avoid duplicate conversations.

However, it seems too early to move the main repository from GitHub. The main benefit I find is the website hosting. According to SimilarWeb, Dark Reader's website has around 1,500,000 monthly views, while Codeberg itself has 400,000. That's a huge bandwidth, works stable, very convenient to use, and we get it all for free (I pay around $15 for some Pro account, but normally it would cost $100-$1000/mo). I don't see any alternative at all, all those AWS stuff is very clumsy, inconvenient and expensive.

I don't see an issue with GitHub or Gitlab being proprietary. I'm ready to pay for the great service. My opinion on Copilot is, if you have really valuable code, which is hard to reproduce, and you expect to benefit from it, never show it. If it's some basic sorting algorithm, I don't see any issue with AI returning it. If it's something more complex, say Dark Reader's open source code, the whole bunch of interconnected functions and modules, the thing should display the MIT License. We can't stop the progress.
